File: pocket-random.py
#!/usr/bin/env python
# pocket-random.py -- randomly pick up some items from Pocket

# Copyright (c) 2016 Shao-Chung Chen
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import json
import os
import random
import sys
import time
import webbrowser

import requests

# API configurations
API_BASE = 'https://getpocket.com/v3/'

class UserConfig(object):
    def __init__(self):
        self.filename = '.pocketrandom'
        self.pathname = os.path.join(os.path.expanduser('~'), self.filename)
        self.api_key = None
        self.username = None
        self.user_code = None
        self.user_token = None
        self.load_or_create_new()

    def load_or_create_new(self):
        try:
            with open(self.pathname, 'rb') as cfgfile:
                cfg = json.loads(cfgfile.read())
                self.api_key = cfg.get('api_key')
                self.username = cfg.get('username')
                self.user_code = cfg.get('user_code')
                self.user_token = cfg.get('user_token')
        except (IOError, ValueError):
            # config file not existing or invalid, create new/empty
            self.save()

    def save(self):
        with open(self.pathname, 'wb') as cfgfile:
            cfgfile.write(json.dumps({
                'api_key': self.api_key,
                'username': self.username,
                'user_code': self.user_code,
                'user_token': self.user_token}))


if __name__ == '__main__':
    # config
    cfg = UserConfig()

    # API key
    if not cfg.api_key:
        pocket_devsite = 'http://getpocket.com/developer/apps/'
        print('No API key available. Get one on {devapps}'.format(devapps=pocket_devsite))
        webbrowser.open(pocket_devsite)
        cfg.api_key = raw_input('Enter your API key: ').strip()
        cfg.save()

    # OAuth authentication
    if not cfg.user_code and not cfg.user_token:
        redirect_uri = 'https://getpocket.com/connected_accounts'
        request_url = '{api_base}oauth/request'.format(api_base=API_BASE)
        request_data = {
            'consumer_key': cfg.api_key,
            'redirect_uri': redirect_uri}
        request_headers = {
            'X-Accept': 'application/json'}

        response = requests.post(request_url, data=request_data, headers=request_headers)
        if response.status_code != requests.codes.ok:
            raise Exception('request code error, status_code=[{status_code}], X-Error-Code=[{xerror}]'.format(
                status_code=response.status_code, xerror=response.headers.get('X-Error-Code')))

        cfg.user_code = response.json()['code']
        print('OAuth code: {oauth_code}'.format(oauth_code=cfg.user_code))
        authorization_url = 'https://getpocket.com/auth/authorize?request_token={request_token}&redirect_uri={redirect_uri}'.format(
            request_token=cfg.user_code, redirect_uri=redirect_uri)
        print('Please authroize this app on {auth_url}'.format(auth_url=authorization_url))
        webbrowser.open(authorization_url)
        raw_input('Press any key to continue...')
        cfg.save()

    if not cfg.user_token:
        request_url = '{api_base}oauth/authorize'.format(api_base=API_BASE)
        request_data = {
            'consumer_key': cfg.api_key,
            'code': cfg.user_code}
        request_headers = {
            'X-Accept': 'application/json'}
        response = requests.post(request_url, data=request_data, headers=request_headers)
        if response.status_code != requests.codes.ok:
            raise Exception('request authorization error, status_code=[{status_code}], X-Error-Code=[{xerror}]'.format(
                status_code=response.status_code, xerror=response.headers.get('X-Error-Code')))

        cfg.username = response.json()['username']
        cfg.user_token = response.json()['access_token']
        print('Username: {username}'.format(username=cfg.username))
        print('Auth token: {oauth_token}'.format(oauth_token=cfg.user_token))
        cfg.save()

    if cfg.api_key and cfg.user_code and cfg.user_token:
        print('Hello {username}!'.format(username=cfg.username))
        print('')
        print('Retrieving items from Pocket...')
        # FIXME: retrieve items in small batches (API limitation = 5000 items for each requests)
        # FIXME: save the retrieved data if possible
        request_url = '{api_base}get'.format(api_base=API_BASE)
        request_data = {
            'consumer_key': cfg.api_key,
            'access_token': cfg.user_token,
            'detailType': 'simple'}
        response = requests.post(request_url, data=request_data)
        items = response.json()['list'].values()
        item_count = len(items)
        print('{item_count} items retrieved!'.format(item_count=item_count))

        while True:
            picked_item = random.choice(items)
            item_id = picked_item.get('item_id')
            item_title = picked_item.get('resolved_title').encode('utf8')
            item_url = picked_item.get('resolved_url').encode('utf8')
            print('')
            print('Item #{item_id} - "{item_title}"({item_url})'.format(
                item_id=item_id, item_title=item_title, item_url=item_url))

            while True:
                answer = raw_input('Action? open(o), archive(a), next(n), quit(q)? ').lower()
                if answer in ['o', 'open']:
                    webbrowser.open(item_url)
                elif answer in ['n', 'next']:
                    break
                elif answer in ['q', 'quit']:
                    sys.exit(0)
                elif answer in ['a', 'archive']:
                    request_url = '{api_base}send'.format(api_base=API_BASE)
                    request_data = {
                        'consumer_key': cfg.api_key,
                        'access_token': cfg.user_token,
                        'actions': json.dumps([{
                            'action': 'archive',
                            'item_id': item_id,
                            'time': int(time.time())}])}
                    response = requests.get(request_url, params=request_data)
                    if response.status_code != requests.codes.ok:
                        raise Exception('archive item error, status_code=[{status_code}], X-Error-Code=[{xerror}]'.format(
                            status_code=response.status_code, xerror=response.headers.get('X-Error-Code')))
                    else:
                        print('Item #{item_id} archived'.format(item_id=item_id))
                    break

File: pocket-random.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"github.com/fatih/color"
	"github.com/toqueteos/webbrowser"
	"golang.org/x/crypto/ssh/terminal"
	"io/ioutil"
	"log"
	"math/rand"
	"net/http"
	"net/url"
	"os"
	"os/user"
	"path"
	"strconv"
	"strings"
	"time"
)

const (
	API_BASE = "https://getpocket.com/v3/"
)

////////////////////////////////////////////////////////////////////////////////

func prettyDateSince(unixTime int) string {
	then := time.Unix(int64(unixTime), 0)
	sinceThen := time.Since(then)
	diffDays := int(sinceThen.Hours() / 24)
	switch {
	case diffDays < 0:
		return "from the future!"
	case diffDays == 0:
		diffSeconds := int(sinceThen.Seconds())
		switch {
		case diffSeconds < 60:
			return "few seconds ago"
		case diffSeconds < 120:
			return "a minute ago"
		case diffSeconds < 3600:
			return fmt.Sprintf("%d minutes ago", int(diffSeconds/60))
		case diffSeconds < 7200:
			return "an hour ago"
		default:
			return fmt.Sprintf("%d hours ago", int(diffSeconds/3600))
		}
	case diffDays == 1:
		return "yesterday"
	case diffDays < 7:
		return fmt.Sprintf("%d days ago", int(diffDays))
	case diffDays < 31:
		return fmt.Sprintf("%d weeks ago", int(diffDays/7))
	case diffDays < 365:
		return fmt.Sprintf("%d months ago", int(diffDays/30))
	default:
		return fmt.Sprintf("%d years ago", int(diffDays/365))
	}
	return ""
}

func truncateString(input string, width int) string {
	if len(input) > width {
		return input[:width-1] + "…"
	}
	return input
}

////////////////////////////////////////////////////////////////////////////////

type UserConfig struct {
	ApiKey    string `json:"api_key"`
	Username  string `json:"username"`
	UserCode  string `json:"user_code"`
	UserToken string `json:"user_token"`
	pathName  string
}

func NewUserConfig() *UserConfig {
	cfgFileName := ".pocketrandom"

	currentUser, err := user.Current()
	if err != nil {
		log.Fatal(err)
	}

	cfg := new(UserConfig)
	cfg.pathName = path.Join(currentUser.HomeDir, cfgFileName)
	return cfg
}

func (cfg *UserConfig) save() {
	cfgJson, err := json.Marshal(cfg)
	if err != nil {
		log.Fatal(err)
	}

	err = ioutil.WriteFile(cfg.pathName, cfgJson, 0644)
	if err != nil {
		log.Fatal(err)
	}
}

func (cfg *UserConfig) loadOrInitialize() *UserConfig {
	raw, err := ioutil.ReadFile(cfg.pathName)
	if err != nil {
		cfg.save()
	}

	err = json.Unmarshal(raw, &cfg)
	if err != nil {
		cfg.save()
	}
	return cfg
}

////////////////////////////////////////////////////////////////////////////////

func requestPocketApi(cfg *UserConfig, requestPath string, requestData url.Values) map[string]interface{} {
	if cfg.ApiKey != "" {
		requestData.Add("consumer_key", cfg.ApiKey)
	}
	if cfg.UserToken != "" {
		requestData.Add("access_token", cfg.UserToken)
	}

	requestUrl := API_BASE + requestPath
	req, err := http.NewRequest("POST", requestUrl, strings.NewReader(requestData.Encode()))
	if err != nil {
		log.Fatal(err)
	}
	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Add("X-Accept", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if (err != nil) || (resp.StatusCode != 200) {
		log.Fatalf("POST '%s' failed, StatusCode=[%d], X-Error-Code=[%s]",
			requestUrl, resp.StatusCode, resp.Header.Get("X-Error-Code"))
	}
	defer resp.Body.Close()

	var bodyJson map[string]interface{}
	err = json.NewDecoder(resp.Body).Decode(&bodyJson)
	if err != nil {
		log.Fatal(err)
	}
	return bodyJson
}

func requestApiKey(cfg *UserConfig) {
	pocketDevSite := "http://getpocket.com/developer/apps/"
	fmt.Printf("No API key available. Get one on %s\n", pocketDevSite)
	webbrowser.Open(pocketDevSite)

	fmt.Printf("Enter your API key: ")
	apiKey, err := bufio.NewReader(os.Stdin).ReadString('\n')
	if err != nil {
		log.Fatal(err)
	}

	cfg.ApiKey = strings.Trim(apiKey, " \t\n")
	cfg.save()
}

func requestUserCode(cfg *UserConfig) {
	const redirectUri = "https://getpocket.com/connected_accounts"
	requestData := url.Values{
		"redirect_uri": {redirectUri},
	}

	body := requestPocketApi(cfg, "oauth/request", requestData)
	userCode := body["code"].(string)
	fmt.Printf("OAuth code: %s\n", userCode)

	authorizationUrl := fmt.Sprintf(
		"https://getpocket.com/auth/authorize?request_token=%s&redirect_uri=%s",
		userCode, redirectUri)
	fmt.Printf("Please authorize this app on %s\n", authorizationUrl)
	webbrowser.Open(authorizationUrl)

	fmt.Printf("Press any key to continue...")
	bufio.NewReader(os.Stdin).ReadBytes('\n')

	cfg.UserCode = userCode
	cfg.save()
}

func requestUserToken(cfg *UserConfig) {
	requestData := url.Values{
		"code": {cfg.UserCode},
	}

	body := requestPocketApi(cfg, "oauth/authorize", requestData)
	userName := body["username"].(string)
	userToken := body["access_token"].(string)
	fmt.Printf("Username: %s\n", userName)
	fmt.Printf("Auth token: %s\n", userToken)

	cfg.Username = userName
	cfg.UserToken = userToken
	cfg.save()
}

func requestPermission(cfg *UserConfig) {
	if cfg.ApiKey == "" {
		requestApiKey(cfg)
	}

	if cfg.UserCode == "" {
		requestUserCode(cfg)
	}

	if cfg.UserToken == "" {
		requestUserToken(cfg)
	}
}

func retrieveItems(cfg *UserConfig) []map[string]interface{} {
	respChnl := make(chan map[string]interface{}, 0)

	go func() {
		requestData := url.Values{
			"detailType": {"simple"},
		}
		body := requestPocketApi(cfg, "get", requestData)
		respChnl <- body
	}()

	// wait for items, and animate the waiting message
	var body map[string]interface{}
	func() {
		for true {
			for _, slash := range "|/-\\" {
				fmt.Printf("Retrieving items from Pocket ... %c\r", slash)
				select {
				case body = <-respChnl:
					return
				case <-time.After(250 * time.Millisecond):
					continue
				}
			}
		}
	}()

	var items []map[string]interface{}
	for _, item := range body["list"].(map[string]interface{}) {
		items = append(items, item.(map[string]interface{}))
	}
	fmt.Printf("Retrieving items from Pocket... %d items retrieved!\n", len(items))

	// shuffle by randomly swap items
	shuffleRounds := rand.Intn(10)
	for sr := 0; sr < shuffleRounds; sr++ {
		for i := range items {
			j := rand.Intn(len(items))
			items[i], items[j] = items[j], items[i]
		}
	}
	return items
}

func userInteractOnItem(cfg *UserConfig, item map[string]interface{}) {
	for true {
		fmt.Printf("Action[help(h), o, a, f, d, n, q] > ")
		userAction, err := bufio.NewReader(os.Stdin).ReadString('\n')
		if err != nil {
			log.Fatal(err)
			os.Exit(1)
		}
		userAction = strings.Trim(strings.ToLower(userAction), " \r\n")

		itemAction := func(action string) {
			actions, err := json.Marshal([]map[string]string{
				{"action": action, "item_id": item["item_id"].(string)},
			})
			if err != nil {
				log.Fatal(err)
			}

			requestData := url.Values{
				"actions": {string(actions)},
			}
			requestPocketApi(cfg, "send", requestData)
			fmt.Printf("Item #%s %sd :-)\n", item["item_id"], action)
		}

		switch userAction {
		case "h", "help":
			fmt.Println(`
help (h) — show this message
open (o) — open the item in web browser
archive (o) — archive the item
favorite (f) — favorite the item
delete (d) — delete the item
next (n) — move to next item
quit (q) — quit this program`)
		case "o", "open":
			webbrowser.Open(item["resolved_url"].(string))
		case "a", "archive":
			itemAction("archive")
			return
		case "f", "favorite":
			itemAction("favorite")
		case "d", "delete":
			itemAction("delete")
			return
		case "n", "next":
			return
		case "q", "quit":
			os.Exit(0)
		}
	}
}

////////////////////////////////////////////////////////////////////////////////

func main() {
	cfg := NewUserConfig().loadOrInitialize()
	requestPermission(cfg)

	// retrieve and go through items
	if cfg.ApiKey == "" || cfg.UserCode == "" || cfg.UserToken == "" {
		log.Fatalf("Configuration error, api_key=[%s], user_code=[%s], user_token=[%s]",
			cfg.ApiKey, cfg.UserCode, cfg.UserToken)
	}

	fmt.Printf("Hello %s!\n", cfg.Username)

	items := retrieveItems(cfg)
	for i := range items {
		item := items[i]
		itemUnixTime, _ := strconv.Atoi(item["time_added"].(string))

		termWidth, _, err := terminal.GetSize(int(os.Stdout.Fd()))
		if err != nil {
			termWidth = 80 // default fallback
		}

		var itemTitle string
		switch {
		case item["resolved_title"] != nil && item["resolved_title"] != "":
			itemTitle = item["resolved_title"].(string)
		case item["given_title"] != nil && item["given_title"] != "":
			itemTitle = item["given_title"].(string)
		}

		itemId := fmt.Sprintf("[#%s]", item["item_id"])
		itemTitle = fmt.Sprintf("\"%s\"", truncateString(itemTitle, termWidth-len("\"\"")-len(itemId)-1))
		itemUrl := truncateString(item["resolved_url"].(string), termWidth-1)
		itemDate := fmt.Sprintf("Added %s", prettyDateSince(itemUnixTime))
		itemWordCount := fmt.Sprintf("~ %s words", item["word_count"])

		// use `fmt.Fprintln(color.Output, …)` to support Windows
		fmt.Println()
		if item["favorite"] != "0" {
			fmt.Fprintln(color.Output, color.RedString("★ FAVORITED"))
		}
		fmt.Fprintln(color.Output, color.YellowString(itemId)+" "+color.WhiteString(itemTitle))
		fmt.Fprintln(color.Output, color.GreenString(strings.Replace(itemUrl, "%", "%%", -1))) // `itemUrl` may contain '%xx' which can be considered as a format string
		fmt.Fprintln(color.Output, color.CyanString(itemDate)+" | "+color.MagentaString(itemWordCount))

		userInteractOnItem(cfg, item)
	}
}
